[Substitution Map] Handle substitutions of generic parameters made concrete

DougGregor opened the pull request:
If `SubstitutionMap` is asked to form a substitution for a generic
parameter that has been made concrete by the generic signature,
substitute into the concrete type. This allows us to better deal with
non-canonical types.

As an added bonus, flatten the storage of replacement types in `SubstitutionMap`. It was a `DenseMap`, now make it an array.

DougGregor committed:
[Substitution Map] Handle substitutions of generic parameters made concrete.

If SubstitutionMap is asked to form a substitution for a generic
parameter that has been made concrete by the generic signature,
substitute into the concrete type. This allows us to better deal with
non-canonical types.

DougGregor commented:
@swift-ci please smoke test and merge

slavapestov commented on lib/AST/SubstitutionMap.cpp:
```diff
@@ -70,10 +70,19 @@ Type SubstitutionMap::lookupSubstitution(CanSubstitutableType type) const {
   if (known != subMap.end() && known->second)
     return known->second;
 
+  // The generic parameter may have been made concrete by the generic signature,
```
Why don't we just put the concrete type in the replacement map when we build the signature?

DougGregor replied:
That would fit better with a flat array of replacement types, I guess.

slavapestov replied:
Yeah.

DougGregor replied:
We have to do it lazily because the concrete types can end up referring to other generic type parameters. Which also means I need to deal with the recursion in ill-formed programs (hence the failures).

DougGregor committed:
[AST] Use flat array storage for SubstitutionMap's replacement types.

The replacement types in a SubstitutionMap correspond with the generic
parameters of its generic signature, so replace the DenseMap storage
with a flat array of Types, one element for each generic parameter.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please clean smoke test Linux

DougGregor commented:
@swift-ci please clean smoke test Linux
